crypto/src/memo.rs

```diff
@@ -16,12 +23,107 @@ impl Default for MemoPlaintext {
 }
 
 impl MemoPlaintext {
-    pub fn encrypt(&self, _key: &Address) -> MemoCiphertext {
-        // TODO!
-        let memo_bytes = [0u8; MEMO_CIPHERTEXT_LEN_BYTES];
-        MemoCiphertext(memo_bytes)
+    /// Encrypt a memo, returning its ciphertext.
+    pub fn encrypt(
+        &self,
+        esk: &ka::Secret,
+        transmission_key: &ka::Public,
+        diversified_generator: &decaf377::Element,
+    ) -> Result<MemoCiphertext, anyhow::Error> {
+        let epk = esk.diversified_public(diversified_generator);
+        let shared_secret = esk
+            .key_agreement_with(&transmission_key)
+            .map_err(|_| anyhow!("could not perform key agreement"))?;
+
+        // Use Blake2b-256 to derive encryption key.
+        let mut kdf_params = blake2b_simd::Params::new();
+        kdf_params.hash_length(32);
+        let mut kdf = kdf_params.to_state();
+        kdf.update(&shared_secret.0);
+        kdf.update(&epk.0);
+        let kdf_output = kdf.finalize();
+        let key = Key::from_slice(kdf_output.as_bytes());
+
+        let cipher = ChaCha20Poly1305::new(key);
+        let nonce = Nonce::from_slice(&[1u8; 12]);
+
+        let encryption_result = cipher
+            .encrypt(nonce, self.0.as_ref())
+            .map_err(|_| anyhow!("encryption error!"))?;
```

Comment:
Could this also be infallible? I looked at the docs and it seems like it's a generic method, maybe the error is there only generically? https://docs.rs/aead/0.4.3/aead/trait.Aead.html#tymethod.encrypt

Reply:
It looks like the generic implementation here https://docs.rs/aead/0.4.3/src/aead/lib.rs.html#360-370 creates a new buffer and then calls `encrypt_in_place` https://docs.rs/aead/0.4.3/aead/trait.AeadInPlace.html#method.encrypt_in_place which errors only if the buffer is wrongly sized.

So the method is infallible, but you can't tell from the types because there's always a generic `Error` type.  It'd be neat if there were a way for the generic impl to specify `Error = !`, but it looks like the crate hardcodes an opaque error type in an attempt to prevent error oracles.

Reply:
So, we could safely `.unwrap()` or `.expect()` this away, and combined with the other two comments, we could make the whole function infallible.